### PR TITLE
Px special inputs

### LINKED ISSUE #2119: px.Constant and the "identity" problem
### `px.Constant`

In certain cases to get the kind of output one would want with PX we need to create a dummy column containing the same string or number in each row. I propose to create a special kind of object that we could pass into the `px` functions to do this like:

```python
px.bar(df, y="the_y_value", x=px.Constant(1, label="the_optional_label"))
```

which would be a shortcut for:

```python
px.bar(df, y="the_y_value", x=[1]*len(df), labels={"x": "the_optional_label"})
```

### `px.Identity` + `px.Direct` (identity option 1)

Relatedly, but on a different note, one might have a column that directly encodes something like the color, and we might want a similar mechanism for doing something like

```python
px.bar(x=[1,2,3], y=[1,2,3], color=px.Identity(["red", "green", "blue"]))
px.bar(x=[1,2,3], y=[1,2,3], color=px.Identity("column_with_colors"))
```

which would be a shortcut for:

```python
px.bar(x=[1,2,3], y=[1,2,3], color=["red", "green", "blue"], 
color_discrete_map={"red":"red", "blue":"blue", "green": "green"})
```

Using the two together would allow you to force a single series to a single color:

```python
px.bar(x=[1,2,3], y=[1,2,3], color=px.Identity(px.Constant("red")))
```

and this might warrant its own shortcut, maybe `px.Direct`?

```python
px.bar(x=[1,2,3], y=[1,2,3], color=px.Direct("red"))
```

which would be a more-intuitive shortcut for the following equivalent calls, in terms of their output:

```python
px.bar(x=[1,2,3], y=[1,2,3], color=px.Constant("red"),
    color_discrete_map={"red":"red"})
px.bar(x=[1,2,3], y=[1,2,3], color=["red"]*3, 
    color_discrete_map={"red":"red"})
px.bar(x=[1,2,3], y=[1,2,3], 
    color_discrete_sequence=["red"])
```

### `px.IDENTITY_MAP` (identity option 2)

A slightly different way of solving the "identity" problem would be to be able to pass in a magic value to `*_map` (edit: which per @emmanuelle could be just the string `"identity"`!) like this:

```python
px.bar(x=[1,2,3], y=[1,2,3], color=["red", "green", "blue"],  
color_discrete_map=px.IDENTITY_MAP)
```

which would also be the equivalent of

```python
px.bar(x=[1,2,3], y=[1,2,3], color=["red", "green", "blue"], 
color_discrete_map={"red":"red", "blue":"blue", "green": "green"})
```

This would avoid the need for nesting `px.Constant` and `px.Identity` when you want to use constant and identity together, and would mean no need for `px.Direct` as shortcut for that:

```python
px.bar(x=[1,2,3], y=[1,2,3], color=px.Constant("red"),  
color_discrete_map=px.IDENTITY_MAP)
```

(note that the foregoing could also apply to any groupable attrs beyond `color` like `symbol` or `line_dash`)

### PR DESCRIPTION
**Pull request by nicolaskruchten**
Closes #2119 ... @emmanuelle LMK what you think of the API and testing approach and if it makes sense I'll document it.

### CODE REVIEW

File: packages/python/plotly/plotly/express/_core.py [hunk 1]
```diff
@@ -41,6 +41,30 @@ def __init__(self):
 defaults = PxDefaults()
 del PxDefaults
 
+
+class IdentityMap(object):
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/express/_core.py]: could these two classes be moved to a `_special_inputs.py` file to shorten a bit `_core.py`?

**Comment by emmanuelle** [comment]
Thanks for the PR! I'm completely convinced by `px.Constant`, a bit less by `px.IdentityMap` which looks like a "dummy" class (it does not do much from what I understand of the code and tests). I'm curious to know why you chose a class instead of the `'identity'` string which we had mentioned. 

I'll do more QA on examples tomorrow, tonight I only had time for a quick pass reading the code.

File: packages/python/plotly/plotly/express/_core.py (hunk 1 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: sure, if needed

**Comment by nicolaskruchten** [comment]
So `IdentityMap` isn't a dummy class, it's the actual implementation: each `IdentityMap` object returns the key on lookup, which is what we want in this case. I didn't implement the `"identity"` string thing yet just to see what this would feel like. I actually like this, personally, as it's a bit more Pythonic in a way, and doesn't have a "magic string" variable in a place where otherwise only a `dict`-like object is accepted. It's pretty trivial to implement the `"identity"` thing as well, just an `if` in `make_mapping()` somwhere to force `val_map` to be an `IdentityMap` when that string is encountered.

File: packages/python/plotly/plotly/express/_core.py (hunk 1 above)
Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_core.py]: as you wish, I just feel that the length of this file makes it a bit overwhelming.

**Comment by emmanuelle** [comment]
Ah ok thanks for the explanations, I had not understood how the class worked. Maybe add a line with your explanation above in the docstring of the class?

**Comment by emmanuelle** [comment]
So I've experimented a bit with your branch. It works well as expected. From a pure usability point of view, doing something like `fig = px.scatter(x=[1, 2], y=[3, 4], color=px.IdentityMap(['yellow', 'blue'])` would be slightly nicer than `fig = px.scatter(x=[1, 2], y=[3, 4], color=['yellow', 'blue'], color_discrete_map=px.IdentityMap())` but on the other hand the current implementation is simple and elegant (so simple I had originally thought that something was missing!). Thoughts?

**Comment by nicolaskruchten** [comment]
Yep, `fig = px.scatter(x=[1, 2], y=[3, 4], color=px.IdentityMap(['yellow', 'blue'])` was one of the options I had listed in #2119 (as `px.Direct`) but after our conversation I thought that the second option seemed better...

**Comment by nicolaskruchten** [comment]
OK I addressed your feedback, including the `"identity"` shortcut.

File: packages/python/plotly/plotly/express/_core.py (hunk 1 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: done!

**Comment by emmanuelle** [comment]
> Yep, `fig = px.scatter(x=[1, 2], y=[3, 4], color=px.IdentityMap(['yellow', 'blue'])` was one of the options I had listed in #2119 (as `px.Direct`) but after our conversation I thought that the second option seemed better...

Oops... reading again our convo in #2119 I think I had preferred option 2 because it used a constant instead of a function or a class, but eventually it's not a constant...

**Comment by nicolaskruchten** [comment]
> I had preferred option 2 because it used a constant instead of a function or a class, but eventually it's not a constant...

I added this in: you can use `"identity"` now also (under the hood it just swaps out an `IdentityMap` for you so you don't have to see it).

I can implement `px.Identity` and `px.Direct` still if you think it will help?

**Comment by emmanuelle** [comment]
Thanks for implementing the `'identity'` thing. `px.Direct` could be useful indeed - I personally prefer using `fig.update_traces` but I understand that a fraction of users will want to do everything from `px`. As for `px.Identity` this is less necessary in my opinion, since when there are more than one colour people they will often exist already as a column in a DataFrame so people might as well do `color='my_color_column', color_discrete_map="identity"`. 

`px.Direct` can also come in another PR so feel free to merge this when you want. About docs: do you plan to add docs as part of this PR or together with one of the WIP PRs?

**Comment by nicolaskruchten** [comment]
I was going to document everything in the last PR :)

**Comment by nicolaskruchten** [comment]
Superseded by #2336

**Pull request closed without merging**